{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = open('trio_example.vcf', 'r')\n",
    "chrom_length = open('chromosome_lengths.tsv', 'r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 394286, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 150602, 0, 0, 13798, 0]\n",
      "[0, 0, 0, 0, 0, 0, 276633, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 107452, 0, 0, 9838, 0]\n",
      "[0, 0, 0, 0, 0, 0, 278793, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 105042, 0, 0, 9684, 0]\n",
      "[0, 0, 0, 0, 0, 0, 277577, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 105865, 0, 0, 10007, 0]\n",
      "~\n",
      "[0, 0, 0, 0, 0, 0, 245630, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 95489, 0, 0, 8401, 0]\n",
      "[0, 0, 0, 0, 0, 0, 247600, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 93270, 0, 0, 8214, 0]\n",
      "[0, 0, 0, 0, 0, 0, 246455, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 94076, 0, 0, 8545, 0]\n",
      "~\n",
      "[0, 0, 0, 0, 0, 0, 31003, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 11963, 0, 0, 1437, 0]\n",
      "[0, 0, 0, 0, 0, 0, 31193, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 11772, 0, 0, 1470, 0]\n",
      "[0, 0, 0, 0, 0, 0, 31122, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 11789, 0, 0, 1462, 0]\n",
      "~\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0015414885947572707, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0014817322411817349, 0.0, 0.0, 5.3838450490815247e-05, 0.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0015538516307531663, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0014472993343214444, 0.0, 0.0, 5.2640046700578074e-05, 0.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0015466660083088514, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0014598062847177465, 0.0, 0.0, 5.4761285495062046e-05, 0.0]\n",
      "~\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0001945640634420049, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0001856335578051618, 0.0, 0.0, 9.209124313212892e-06, 0.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0001957564374720659, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00018266975194201828, 0.0, 0.0, 9.42060733501945e-06, 0.0]\n",
      "[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00019531086612399048, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00018293354618114624, 0.0, 0.0, 9.369338723672406e-06, 0.0]\n"
     ]
    }
   ],
   "source": [
    "chrom_number = [\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\",\"13\",\"14\",\"15\",\"16\",\"17\",\"18\",\"19\",\"20\",\"21\",\"22\",\"X\",\"Y\"]\n",
    "index = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24]\n",
    "\n",
    "mutations_count = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "\n",
    "total_count_N1 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "total_count_N2 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "total_count_N3 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "\n",
    "SNVs_count_N1 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "SNVs_count_N2 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "SNVs_count_N3 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "\n",
    "Indels_N1 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "Indels_N2 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "Indels_N3 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "\n",
    "Freq_N1_Indels = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "Freq_N2_Indels = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "Freq_N3_Indels = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "\n",
    "Freq_N1_SNVs = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "Freq_N2_SNVs = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "Freq_N3_SNVs = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "\n",
    "for line in vcf_file:\n",
    "    \n",
    "    # check if the line is an info line, starting with ##, in that case, skip\n",
    "    if (line.startswith(\"##\")):\n",
    "        continue\n",
    "    if (line.startswith(\"#\")):\n",
    "        #here we want to find the index of each column\n",
    "        header=line.split(\"\\t\") # split by tabulator\n",
    "        \n",
    "        for item in enumerate(header):\n",
    "            if (item[1] == 'N1'):\n",
    "                index_N1=item[0]\n",
    "            if (item[1] == 'N2'):\n",
    "                index_N2=item[0]\n",
    "            if (item[1] == 'N3\\n'):\n",
    "                index_N3=item[0]\n",
    "            if (item[1] == \"#CHROM\"):\n",
    "                index_chrom = item[0]\n",
    "    else:           \n",
    "        info = line.split(\"\\t\")\n",
    "        N1=info[index_N1]\n",
    "        N2=info[index_N2]\n",
    "        N3=info[index_N3].strip()  # we use strip function to get rid of \"\\n\" at the end of each line.\n",
    "        chrom = info[index_chrom]\n",
    "        \n",
    "        \n",
    "        for i in chrom_number:\n",
    "            if (i == chrom):\n",
    "                if (i == \"X\"):\n",
    "                    mutations_count[22] = mutations_count[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    mutations_count[23] = mutations_count[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    mutations_count[x -1] = mutations_count[x-1] + 1 \n",
    "                \n",
    "        \n",
    "        # after printing the results, we can see that the mutations are only on chromosome 7, 20, and \"X\".\n",
    "        #######\n",
    "        \n",
    "        # now we count them for each indiviual\n",
    "        \n",
    "        \n",
    "        \n",
    "        # N1 total number of mutations on each chromosome\n",
    "        for i in chrom_number:\n",
    "            if (i == chrom and N1 != \"0/0\"):\n",
    "                if (i == \"X\"):\n",
    "                    total_count_N1[22] = total_count_N1[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    total_count_N1[23] = total_count_N1[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    total_count_N1[x -1] = total_count_N1[x-1] + 1 \n",
    "                    \n",
    "        # N2 total number of mutations on each chromosome          \n",
    "        for i in chrom_number:\n",
    "            if (i == chrom and N2 != \"0/0\"):\n",
    "                if (i == \"X\"):\n",
    "                    total_count_N2[22] = total_count_N2[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    total_count_N2[23] = total_count_N2[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    total_count_N2[x -1] = total_count_N2[x-1] + 1 \n",
    "                    \n",
    "        # N3 total number of mutations on each chromosome           \n",
    "        for i in chrom_number:\n",
    "            if (i == chrom and N3 != \"0/0\"):\n",
    "                if (i == \"X\"):\n",
    "                    total_count_N3[22] = total_count_N3[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    total_count_N3[23] = total_count_N3[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    total_count_N3[x -1] = total_count_N3[x-1] + 1 \n",
    "                    \n",
    "                    \n",
    "       # now we only count SNVs for each indiviual, after that we count Indels. \n",
    "        # N1 SNVs\n",
    "        for i in chrom_number:\n",
    "            if (i == chrom and N1 != \"0/0\"):\n",
    "                if (len(info[3])!=1):\n",
    "                    continue\n",
    "                if (len(info[4])!=1):\n",
    "                    continue\n",
    "                if (i == \"X\"):\n",
    "                    SNVs_count_N1[22] = SNVs_count_N1[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    SNVs_count_N1[23] = SNVs_count_N1[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    SNVs_count_N1[x -1] = SNVs_count_N1[x-1] + 1 \n",
    "                    \n",
    "        # N2 SNVs           \n",
    "        for i in chrom_number:\n",
    "            if (i == chrom and N2 != \"0/0\"):\n",
    "                if (len(info[3])!=1):\n",
    "                    continue\n",
    "                if (len(info[4])!=1):\n",
    "                    continue\n",
    "                if (i == \"X\"):\n",
    "                    SNVs_count_N2[22] = SNVs_count_N2[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    SNVs_count_N2[23] = SNVs_count_N2[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    SNVs_count_N2[x -1] = SNVs_count_N2[x-1] + 1 \n",
    "                    \n",
    "        # N3 SNVs           \n",
    "        for i in chrom_number:\n",
    "            if (i == chrom and N3 != \"0/0\"):\n",
    "                if (len(info[3])!=1):\n",
    "                    continue\n",
    "                if (len(info[4])!=1):\n",
    "                    continue\n",
    "                if (i == \"X\"):\n",
    "                    SNVs_count_N3[22] = SNVs_count_N3[22] + 1\n",
    "                    continue\n",
    "                if (i == \"Y\"):\n",
    "                    SNVs_count_N3[23] = SNVs_count_N3[23] + 1\n",
    "                    continue\n",
    "                else:\n",
    "                    x = int(i)\n",
    "                    SNVs_count_N3[x -1] = SNVs_count_N3[x-1] + 1 \n",
    "        \n",
    "        \n",
    "        # N1 Indels\n",
    "        for i in index:\n",
    "            Indels_N1[i-1] = total_count_N1[i-1] - SNVs_count_N1[i-1]\n",
    "        \n",
    "        # N2 Indels\n",
    "        for i in index:\n",
    "            Indels_N2[i-1] = total_count_N2[i-1] - SNVs_count_N2[i-1]\n",
    "        \n",
    "        # N3 Indels\n",
    "        for i in index:\n",
    "            Indels_N3[i-1] = total_count_N3[i-1] - SNVs_count_N3[i-1]\n",
    "            \n",
    "     \n",
    "    \n",
    "    \n",
    "\n",
    "# counting the frequencies\n",
    "        \n",
    "    \n",
    "\n",
    "for line2 in chrom_length:\n",
    "    length = line2.split(\"\\t\")\n",
    "    for i in chrom_number:\n",
    "        if (length[0] == i):\n",
    "            if (i == \"X\"):\n",
    "                Freq_N1_SNVs[22] = SNVs_count_N1[22]/int(length[1])\n",
    "                Freq_N2_SNVs[22] = SNVs_count_N2[22]/int(length[1])\n",
    "                Freq_N3_SNVs[22] = SNVs_count_N3[22]/int(length[1])\n",
    "                Freq_N1_Indels[22] = Indels_N1[22]/int(length[1])\n",
    "                Freq_N2_Indels[22] = Indels_N2[22]/int(length[1])\n",
    "                Freq_N3_Indels[22] = Indels_N3[22]/int(length[1])\n",
    "                \n",
    "                continue\n",
    "            if (i == \"Y\"):\n",
    "                Freq_N1_SNVs[23] = SNVs_count_N1[23]/int(length[1])\n",
    "                Freq_N2_SNVs[23] = SNVs_count_N2[23]/int(length[1])\n",
    "                Freq_N3_SNVs[23] = SNVs_count_N3[23]/int(length[1])\n",
    "                Freq_N1_Indels[23] = Indels_N1[23]/int(length[1])\n",
    "                Freq_N2_Indels[23] = Indels_N2[23]/int(length[1])\n",
    "                Freq_N3_Indels[23] = Indels_N3[23]/int(length[1])\n",
    "                continue\n",
    "            else:\n",
    "                x = int(i)\n",
    "                Freq_N1_SNVs[x -1] =  SNVs_count_N1[x-1]/int(length[1])\n",
    "                Freq_N2_SNVs[x -1] =  SNVs_count_N2[x-1]/int(length[1])\n",
    "                Freq_N3_SNVs[x -1] =  SNVs_count_N3[x-1]/int(length[1])\n",
    "                Freq_N1_Indels[x -1] =  Indels_N1[x-1]/int(length[1])\n",
    "                Freq_N2_Indels[x -1] =  Indels_N2[x-1]/int(length[1])\n",
    "                Freq_N3_Indels[x -1] =  Indels_N3[x-1]/int(length[1])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "                \n",
    "print(mutations_count) \n",
    "print (total_count_N1)\n",
    "print (total_count_N2)\n",
    "print (total_count_N3)\n",
    "print (\"~\")       \n",
    "print (SNVs_count_N1)       \n",
    "print (SNVs_count_N2) \n",
    "print (SNVs_count_N3)\n",
    "print(\"~\")\n",
    "print (Indels_N1)\n",
    "print (Indels_N2)\n",
    "print (Indels_N3)\n",
    "\n",
    "print (\"~\")\n",
    "print (Freq_N1_SNVs)\n",
    "print (Freq_N2_SNVs)\n",
    "print (Freq_N3_SNVs)\n",
    "print(\"~\")\n",
    "print (Freq_N1_Indels)\n",
    "print (Freq_N2_Indels)\n",
    "print (Freq_N3_Indels)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 [python/3.8]",
   "language": "python",
   "name": "sys_python38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
