{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = open('trio_example.vcf', 'r')\n",
    "chrom_length = open('chromosome_lengths.tsv', 'r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found the header['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'N1', 'N2', 'N3\\n']\n",
      "total mutations : 558686\n",
      "number of mutations in N1 : 393923\n",
      "number of mutations in N2 : 393519\n",
      "number of mutations in N3 : 393449\n",
      "~\n",
      "total SNVs : 497631\n",
      "SNVs in N1 : 349520\n",
      "SNVs in N2 : 349084\n",
      "SNVs in N3 : 349076\n",
      "~\n",
      "total number of Indels : 61055\n",
      "Indels in N1 : 44403\n",
      "Indels in N2 : 44435\n",
      "Indels in N3 : 44373\n"
     ]
    }
   ],
   "source": [
    "total_mutations = 0\n",
    "nu_mutations_N1 = 0\n",
    "nu_mutations_N2 = 0\n",
    "nu_mutations_N3 = 0\n",
    "\n",
    "total_SNV = 0\n",
    "nu_SNV_N1 = 0\n",
    "nu_SNV_N2 = 0\n",
    "nu_SNV_N3 = 0\n",
    "for line in vcf_file:\n",
    "    \n",
    "    # check if the line is an info line, starting with ##, in that case, skip\n",
    "    if (line.startswith(\"##\")):\n",
    "        continue\n",
    "    if (line.startswith(\"#\")):\n",
    "        #here we want to find the index of each column\n",
    "        header=line.split(\"\\t\") # split by tabulator\n",
    "        print (\"found the header\" + str(header))\n",
    "        for item in enumerate(header):\n",
    "            if (item[1] == 'N1'):\n",
    "                index_N1=item[0]\n",
    "            if (item[1] == 'N2'):\n",
    "                index_N2=item[0]\n",
    "            if (item[1] == 'N3\\n'):\n",
    "                index_N3=item[0]\n",
    "                \n",
    "    #counting how many mutations we have in the file,then we break up the results for each individual          \n",
    "    else:\n",
    "        info = line.split(\"\\t\")\n",
    "        N1=info[index_N1]\n",
    "        N2=info[index_N2]\n",
    "        N3=info[index_N3].strip()  # we use strip function to get rid of \"\\n\" at the end of each line.\n",
    "        \n",
    "        \n",
    "        total_mutations = total_mutations + 1\n",
    "        \n",
    "        if (N1 != \"0/0\"):\n",
    "            nu_mutations_N1 = nu_mutations_N1 + 1\n",
    "            \n",
    "        \n",
    "        if (N2 != \"0/0\"):\n",
    "            nu_mutations_N2 = nu_mutations_N2 + 1\n",
    "            \n",
    "       \n",
    "        if (N3 != \"0/0\"):\n",
    "            nu_mutations_N3 = nu_mutations_N3 + 1\n",
    "            \n",
    "        #now we count how many SNVs and Indels for each indiviual    \n",
    "        \n",
    "        if (len(info[3])!=1):\n",
    "            continue\n",
    "        if (len(info[4])!=1):\n",
    "            continue\n",
    "            \n",
    "        total_SNV = total_SNV + 1\n",
    "        \n",
    "        \n",
    "        if (N1 != \"0/0\"):\n",
    "            nu_SNV_N1 = nu_SNV_N1 + 1\n",
    "        if (N2 != \"0/0\"):\n",
    "            nu_SNV_N2 = nu_SNV_N2 + 1\n",
    "        if (N3 != \"0/0\"):\n",
    "            nu_SNV_N3 = nu_SNV_N3 + 1\n",
    "            \n",
    "           \n",
    "        \n",
    "        \n",
    "        \n",
    "print(\"total mutations : \" + str(total_mutations)) \n",
    "print(\"number of mutations in N1 : \" + str(nu_mutations_N1))\n",
    "print(\"number of mutations in N2 : \" + str(nu_mutations_N2))\n",
    "print(\"number of mutations in N3 : \" + str(nu_mutations_N3))\n",
    "\n",
    "print (\"~\")\n",
    "print(\"total SNVs : \" + str(total_SNV))\n",
    "print(\"SNVs in N1 : \" + str(nu_SNV_N1))\n",
    "print(\"SNVs in N2 : \" + str(nu_SNV_N2))\n",
    "print(\"SNVs in N3 : \" + str(nu_SNV_N3))\n",
    "    \n",
    "print (\"~\")\n",
    "total_Indels = total_mutations - total_SNV \n",
    "N1_Indels = nu_mutations_N1 - nu_SNV_N1\n",
    "N2_Indels = nu_mutations_N2 - nu_SNV_N2\n",
    "N3_Indels = nu_mutations_N3 - nu_SNV_N3\n",
    "\n",
    "print (\"total number of Indels : \" + str(total_Indels))\n",
    "print(\"Indels in N1 : \" + str(N1_Indels)) \n",
    "print(\"Indels in N2 : \" + str(N2_Indels))  \n",
    "print(\"Indels in N3 : \" + str(N3_Indels))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 [python/3.8]",
   "language": "python",
   "name": "sys_python38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
